CalendarProject/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -0,0 +1,128 @@
+
+    /* 
+    * Given the timeRanges in one day, this function returns the list of complementerTimeRanges that are in the same day, but are not present in timeRanges.
+    * The union of timeRanges and complementerTimeRanges is an entire day.
+    */
+    private List<TimeRange> getComplementerTimeRanges(List<TimeRange> timeRanges) {
+        List<TimeRange> complementerTimeRanges = new LinkedList<TimeRange>();
+        int lastRangeEndPoint = 0;
+        for (TimeRange timeRange : timeRanges) {
+             if (lastRangeEndPoint < timeRange.start()) {
+                 complementerTimeRanges.add(TimeRange.fromStartEnd(lastRangeEndPoint, timeRange.start(), false));
+             }
+             lastRangeEndPoint = timeRange.end();
+        }
+        if (lastRangeEndPoint < TimeRange.END_OF_DAY ) {
+            complementerTimeRanges.add(TimeRange.fromStartEnd(lastRangeEndPoint, TimeRange.END_OF_DAY, true));
+        }
+        return complementerTimeRanges;
+    }
+
+    /*
+    * Removes the duplications from the overlapping timeRanges, and merges them into one single TimeRange. 
+    * Returns a list containing a minimal number of timeRanges, which do not overlpa, but they do contain each original timeRange.
+    */
+    private List<TimeRange> getReducedListOfTimeRanges(List<TimeRange> timeRanges) {
+        List<TimeRange> reducedTimeRanges = new LinkedList<>();
+        if (timeRanges.size()==0) {
+            return reducedTimeRanges;
+        }
+        TimeRange expandableTimeRange = timeRanges.get(0);
+        TimeRange currentTimeRange;
+        for (int i = 1; i<timeRanges.size(); i++) {
+            currentTimeRange = timeRanges.get(i);
+            if (expandableTimeRange.overlaps(currentTimeRange)) { 
+                //unite this timeRange with the previous one(s)
+                expandableTimeRange = expandableTimeRange.getUnion(currentTimeRange);
+            }
+            else { 
+                //add the previous timeRange to the list of reduced timeRanges, because it can't be expanded anymore
+                //start a new cuurentTimeRange
+                reducedTimeRanges.add(expandableTimeRange);
+                expandableTimeRange = currentTimeRange;
+            }
+        }
+        reducedTimeRanges.add(expandableTimeRange);
+        return reducedTimeRanges;
+    }
+
+    /* 
+    * creates a list containing the time ranges only out of the events. 
+    */
+    private List<TimeRange> getListOfTimeRanges(Collection<Event> events) {
+        List<TimeRange> result = new ArrayList<>();
+        for (Event event : events) {
+            result.add(event.getWhen());
+        }
+        return result;
+    }
+
+    /* 
+    * Checks if there is exists an intersection(a common attendee) of the attendee list of two events.
+    */
+    private boolean existCommonAttendees(Collection<String> attendeesEventA, Collection<String> attendeesEventB) {
+        //swap collections so that I can iterate through the shorter one
```

Comment:
Why do you think iterate through the shorter one is faster? both input Collections are not Set.

Reply:
The attendees are represented both in the Event and in the MeetingRequetst class by Sets(HashSets, to be concrete), and the getAttendees() function returns an unmodifiable view of them.  Thus, with this implementation, swapping the collections may significantly reduce the running time. 
Of course, my method does not require the input functions to be HashSets necessarily.  I assumed that swapping the pointers is not an expensive operation, so if the implementations of Event and meetingRequest change, and they will return another kind of Collections, the method will still work correctly without any expensive unnecessary operations.